Error in console (Custom Advancements)

Good plugin but I have some errors like:
[00:09:37 ERROR]: [CustomAdvancements] Could not send advancements to .MclovinXD5638!
[00:11:09 ERROR]: [CustomAdvancements] Could not update advancement Jugar.Jugar2 for rodtestrini!
[04:49:18 ERROR]: [CustomAdvancements] Failed to set put back advancement when setting the parent of null of tree Pescar and advancement Pescar1
Any solution?

Thank you, this helps!

On Tue, 28 Mar 2023 at 19:16, MrRoddd ***@***.***> wrote:

> This is the error that apperas ti many achievements when I start my server:
>
> [12:36:28] [Server thread/INFO]: [CustomAdvancements] Enabling
> CustomAdvancements v1.2.7
> [12:36:28] [Server thread/INFO]: [CustomAdvancements] Loading
> CustomAdvancements messages.yml
> [12:36:28] [Server thread/INFO]: [CustomAdvancements] Progress per packet
> is now 5
> [12:36:28] [Server thread/INFO]: [CustomAdvancements] Advancements per
> packet is now 5
> [12:36:28] [Server thread/INFO]: [CustomAdvancements] Attempting to load
> advancement tree Tradear.yml
> [12:36:28] [Server thread/INFO]: [CustomAdvancements] Loaded advancement
> tree Tradear.yml
> [12:36:28] [Server thread/INFO]: [CustomAdvancements] Attempting to load
> advancement tree Pescar.yml
> [12:36:28] [Server thread/INFO]: [CustomAdvancements] Loaded advancement
> tree Pescar.yml
> [12:36:28] [Server thread/INFO]: [CustomAdvancements] Attempting to load
> advancement tree saltar.yml
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar1' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar2' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar3' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar4' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar5' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar6' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar7' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar8' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar9' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar10' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar11' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar12' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar13' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar14' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar15' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar16' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar17' of tree 'saltar' did not have a value!
> [12:36:28] [Server thread/WARN]: [CustomAdvancements] Advancement
> 'caminar18' of tree 'saltar' did not have a value!
>
> Another arror:
>
> [12:36:28] [ForkJoinPool.commonPool-worker-1/ERROR]: [CustomAdvancements]
> Failed to set put back advancement when setting the parent of null of tree
> Jugar and advancement Jugar4
> java.lang.RuntimeException: Failed to set parent of advancement.
> at me.tippie.customadvancements.v1_19_R2.setParent(v1_19_R2.java:408)
> ~[CustomAdvancements-1.2.7 (1).jar:?]
> at
> me.tippie.customadvancements.v1_19_R2.lambda$loadAdvancements$1(v1_19_R2.java:150)
> ~[CustomAdvancements-1.2.7 (1).jar:?]
> at
> java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
> ~[?:?]
> at
> java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
> ~[?:?]
> at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373) ~[?:?]
> at
> java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
> ~[?:?]
> at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655) ~[?:?]
> at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
> ~[?:?]
> at
> java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
> ~[?:?]
> Caused by: java.lang.NullPointerException: Cannot invoke
> "net.minecraft.resources.MinecraftKey.hashCode()" because "this.d" is null
> at net.minecraft.advancements.Advancement.hashCode(Advancement.java:136)
> ~[paper-1.19.3.jar:git-Paper-448]
> at java.util.HashMap.hash(HashMap.java:338) ~[?:?]
> at java.util.HashMap.put(HashMap.java:610) ~[?:?]
> at java.util.HashSet.add(HashSet.java:221) ~[?:?]
> at net.minecraft.advancements.Advancement.addChild(Advancement.java:116)
> ~[paper-1.19.3.jar:git-Paper-448]
> at net.minecraft.advancements.Advancement.(Advancement.java:53)
> ~[paper-1.19.3.jar:git-Paper-448]
> at
> net.minecraft.advancements.Advancement$Builder.build(Advancement.java:259)
> ~[?:?]
> at me.tippie.customadvancements.v1_19_R2.setParent(v1_19_R2.java:400)
> ~[CustomAdvancements-1.2.7 (1).jar:?]
> ... 8 more
> [12:36:28] [ForkJoinPool.commonPool-worker-1/ERROR]: [CustomAdvancements]
> Failed to set put back advancement when setting the parent of null of tree
> Jugar and advancement Jugar11
> java.lang.RuntimeException: Failed to set parent of advancement.
> at me.tippie.customadvancements.v1_19_R2.setParent(v1_19_R2.java:408)
> ~[CustomAdvancements-1.2.7 (1).jar:?]
> at
> me.tippie.customadvancements.v1_19_R2.lambda$loadAdvancements$1(v1_19_R2.java:150)
> ~[CustomAdvancements-1.2.7 (1).jar:?]
> at
> java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
> ~[?:?]
> at
> java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
> ~[?:?]
> at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373) ~[?:?]
> at
> java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
> ~[?:?]
> at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655) ~[?:?]
> at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
> ~[?:?]
> at
> java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
> ~[?:?]
> Caused by: java.lang.NullPointerException: Cannot invoke
> "net.minecraft.resources.MinecraftKey.hashCode()" because "this.d" is null
> at net.minecraft.advancements.Advancement.hashCode(Advancement.java:136)
> ~[paper-1.19.3.jar:git-Paper-448]
> at java.util.HashMap.hash(HashMap.java:338) ~[?:?]
> at java.util.HashMap.put(HashMap.java:610) ~[?:?]
> at java.util.HashSet.add(HashSet.java:221) ~[?:?]
> at net.minecraft.advancements.Advancement.addChild(Advancement.java:116)
> ~[paper-1.19.3.jar:git-Paper-448]
> at net.minecraft.advancements.Advancement.(Advancement.java:53)
> ~[paper-1.19.3.jar:git-Paper-448]
> at
> net.minecraft.advancements.Advancement$Builder.build(Advancement.java:259)
> ~[?:?]
> at me.tippie.customadvancements.v1_19_R2.setParent(v1_19_R2.java:400)
> ~[CustomAdvancements-1.2.7 (1).jar:?]
> ... 8 more
> [12:36:28] [ForkJoinPool.commonPool-worker-1/ERROR]: [CustomAdvancements]
> Failed to set put back advancement when setting the parent of null of tree
> Jugar and advancement Jugar3
> java.lang.RuntimeException: Failed to set parent of advancement.
> at me.tippie.customadvancements.v1_19_R2.setParent(v1_19_R2.java:408)
> ~[CustomAdvancements-1.2.7 (1).jar:?]
> at
> me.tippie.customadvancements.v1_19_R2.lambda$loadAdvancements$1(v1_19_R2.java:150)
> ~[CustomAdvancements-1.2.7 (1).jar:?]
> at
> java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
> ~[?:?]
> at
> java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
> ~[?:?]
> at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373) ~[?:?]
> at
> java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
> ~[?:?]
> at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655) ~[?:?]
> at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
> ~[?:?]
> at
> java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
> ~[?:?]
> Caused by: java.lang.NullPointerException: Cannot invoke
> "net.minecraft.resources.MinecraftKey.hashCode()" because "this.d" is null
> at net.minecraft.advancements.Advancement.hashCode(Advancement.java:136)
> ~[paper-1.19.3.jar:git-Paper-448]
> at java.util.HashMap.hash(HashMap.java:338) ~[?:?]
> at java.util.HashMap.put(HashMap.java:610) ~[?:?]
> at java.util.HashSet.add(HashSet.java:221) ~[?:?]
> at net.minecraft.advancements.Advancement.addChild(Advancement.java:116)
> ~[paper-1.19.3.jar:git-Paper-448]
> at net.minecraft.advancements.Advancement.(Advancement.java:53)
> ~[paper-1.19.3.jar:git-Paper-448]
> at
> net.minecraft.advancements.Advancement$Builder.build(Advancement.java:259)
> ~[?:?]
> at me.tippie.customadvancements.v1_19_R2.setParent(v1_19_R2.java:400)
> ~[CustomAdvancements-1.2.7 (1).jar:?]
> ... 8 more
>
> This last error happens with many more achievements but I have no space.
>
> Error after server starts:
>
> [12:40:45 ERROR]: [CustomAdvancements] Could not send advancements to
> MeRiiToYii!
>
> This hapnes with many more types of achievements.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rowan-vr/CustomAdvancements/issues/118#issuecomment-1487311727>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF3TQRKUUBX6AB3DNQEOF2TW6MMHJANCNFSM6AAAAAAWKYT374>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

[15:58:59 ERROR]: [CustomAdvancements] Could not update advancement Jugar.Jugar2 for Sempiterno_jess!
java.util.concurrent.CompletionException: java.lang.NullPointerException
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320) ~[?:?]
        at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1807) ~[?:?]
        at java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796) ~[?:?]
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373) ~[?:?]
        at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182) ~[?:?]
        at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655) ~[?:?]
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622) ~[?:?]
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165) ~[?:?]
Caused by: java.lang.NullPointerException
        at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:889) ~[guava-31.1-jre.jar:?]
        at com.google.common.collect.SingletonImmutableSet.<init>(SingletonImmutableSet.java:40) ~[guava-31.1-jre.jar:?]
        at com.google.common.collect.ImmutableSet.of(ImmutableSet.java:89) ~[guava-31.1-jre.jar:?]
        at com.google.common.collect.ImmutableSet.construct(ImmutableSet.java:199) ~[guava-31.1-jre.jar:?]
        at com.google.common.collect.ImmutableSet.copyOf(ImmutableSet.java:241) ~[guava-31.1-jre.jar:?]
        at net.minecraft.network.protocol.game.ClientboundUpdateAdvancementsPacket.<init>(ClientboundUpdateAdvancementsPacket.java:30) ~[?:?]
        at me.tippie.customadvancements.v1_19_R2.lambda$sendAdvancementPacketImpl$6(v1_19_R2.java:329) ~[CustomAdvancements-1.2.7 (1).jar:?]
        at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804) ~[?:?]
        ... 6 more

--- a/core/src/main/java/me/tippie/customadvancements/advancement/AdvancementManager.java
+++ b/core/src/main/java/me/tippie/customadvancements/advancement/AdvancementManager.java
@@ -23,202 +23,210 @@ import java.util.logging.Level;
  */
 public class AdvancementManager {
 
-	/**
-	 * Map with key type label and value the {@link AdvancementType} belonging to it.
-	 */
-	private final Map<String, AdvancementType> advancementTypes = new HashMap<>();
-	/**
-	 * Map with key tree label and value the {@link AdvancementTree} belonging to it.
-	 */
-	private final Map<String, AdvancementTree> advancementTrees = new HashMap<>();
-	/**
-	 * Map with key tree label and value the {@link AdvancementRewardType} belonging to it.
-	 */
-	private final Map<String, AdvancementRewardType> advancementRewardTypes = new HashMap<>();
-	/**
-	 * Map with key tree label and value the {@link AdvancementRewardType} belonging to it.
-	 */
-	private final Map<String, AdvancementRequirementType> advancementRequirementTypes = new HashMap<>();
+    /**
+     * Map with key type label and value the {@link AdvancementType} belonging to it.
+     */
+    private final Map<String, AdvancementType> advancementTypes = new HashMap<>();
+    /**
+     * Map with key tree label and value the {@link AdvancementTree} belonging to it.
+     */
+    private final Map<String, AdvancementTree> advancementTrees = new HashMap<>();
+    /**
+     * Map with key tree label and value the {@link AdvancementRewardType} belonging to it.
+     */
+    private final Map<String, AdvancementRewardType> advancementRewardTypes = new HashMap<>();
+    /**
+     * Map with key tree label and value the {@link AdvancementRewardType} belonging to it.
+     */
+    private final Map<String, AdvancementRequirementType> advancementRequirementTypes = new HashMap<>();
 
 
-	/**
-	 * Makes new {@link AdvancementManager}
-	 */
-	public AdvancementManager() {
-	}
+    /**
+     * Makes new {@link AdvancementManager}
+     */
+    public AdvancementManager() {
+    }
 
-	/**
-	 * Registers an advancement type with the plugin, example usage:
-	 * {@code advancementManager.registerAdvancement(new ExampleType());}
-	 *
-	 * @param advancementType the instance of an advancement type
-	 */
-	public void registerAdvancement(final AdvancementType advancementType) {
-		CustomAdvancements.getInstance().getServer().getPluginManager().registerEvents(advancementType, CustomAdvancements.getInstance());
-		advancementTypes.put(advancementType.getLabel(), advancementType);
-	}
+    /**
+     * Registers an advancement type with the plugin, example usage:
+     * {@code advancementManager.registerAdvancement(new ExampleType());}
+     *
+     * @param advancementType the instance of an advancement type
+     */
+    public void registerAdvancement(final AdvancementType advancementType) {
+        CustomAdvancements.getInstance().getServer().getPluginManager().registerEvents(advancementType, CustomAdvancements.getInstance());
+        advancementTypes.put(advancementType.getLabel(), advancementType);
+    }
 
-	/**
-	 * Registers an advancement reward type with the plugin, example usage:
-	 * {@code advancementManager.registerAdvancementReward(new ExampleType());}
-	 *
-	 * @param advancementRewardType the instance of an advancement type
-	 */
-	public void registerAdvancementReward(final AdvancementRewardType advancementRewardType) {
-		advancementRewardTypes.put(advancementRewardType.getLabel(), advancementRewardType);
-	}
+    /**
+     * Registers an advancement reward type with the plugin, example usage:
+     * {@code advancementManager.registerAdvancementReward(new ExampleType());}
+     *
+     * @param advancementRewardType the instance of an advancement type
+     */
+    public void registerAdvancementReward(final AdvancementRewardType advancementRewardType) {
+        advancementRewardTypes.put(advancementRewardType.getLabel(), advancementRewardType);
+    }
 
 
-	/**
-	 * Registers an advancement requirement type with the plugin, example usage:
-	 * {@code advancementManager.registerAdvancementRequirement(new ExampleType());}
-	 *
-	 * @param advancementRequirementType the instance of an advancement type
-	 */
-	public void registerAdvancementRequirement(final AdvancementRequirementType advancementRequirementType) {
-		advancementRequirementTypes.put(advancementRequirementType.getLabel(), advancementRequirementType);
-	}
+    /**
+     * Registers an advancement requirement type with the plugin, example usage:
+     * {@code advancementManager.registerAdvancementRequirement(new ExampleType());}
+     *
+     * @param advancementRequirementType the instance of an advancement type
+     */
+    public void registerAdvancementRequirement(final AdvancementRequirementType advancementRequirementType) {
+        advancementRequirementTypes.put(advancementRequirementType.getLabel(), advancementRequirementType);
+    }
 
-	/**
-	 * Loads the advancement trees and puts them into {@link AdvancementManager#advancementTrees}
-	 */
-	public void loadAdvancements() {
-		advancementTrees.clear();
-		final Path advancementFolder = Paths.get(CustomAdvancements.getInstance().getDataFolder() + "/advancement-trees");
-		if (!Files.exists(advancementFolder)) {
-			try {
-				Files.createDirectories(advancementFolder);
-				CustomAdvancements.getInstance().saveResource("advancement-trees/example.yml", false);
-			} catch (final IOException e) {
-				CustomAdvancements.getInstance().getLogger().log(Level.SEVERE, "Failed to read and/or create plugin directory.",e);
-			}
-		}
-		final File dir = new File(advancementFolder.toString());
-		final File[] advancementDirectoryContent = dir.listFiles();
-		assert advancementDirectoryContent != null;
-		for (final File file : advancementDirectoryContent) {
-			if (file.getName().endsWith(".yml")) {
-				AdvancementTree tree = new AdvancementTree(file);
-				if (file.getName().split(".yml")[0].contains(" ")){
-					CustomAdvancements.getInstance().getLogger().log(Level.SEVERE, "Advancement tree file name '"+file.getName()+"' contains spaces, this is not allowed.");
-					continue;
-				}
-				advancementTrees.put(file.getName().split(".yml")[0], tree);
-			}
-		}
-	}
+    /**
+     * Loads the advancement trees and puts them into {@link AdvancementManager#advancementTrees}
+     */
+    public void loadAdvancements() {
+        advancementTrees.clear();
+        final Path advancementFolder = Paths.get(CustomAdvancements.getInstance().getDataFolder() + "/advancement-trees");
+        if (!Files.exists(advancementFolder)) {
+            try {
+                Files.createDirectories(advancementFolder);
+                CustomAdvancements.getInstance().saveResource("advancement-trees/example.yml", false);
+            } catch (final IOException e) {
+                CustomAdvancements.getInstance().getLogger().log(Level.SEVERE, "Failed to read and/or create plugin directory.", e);
+            }
+        }
+        final File dir = new File(advancementFolder.toString());
+        final File[] advancementDirectoryContent = dir.listFiles();
+        assert advancementDirectoryContent != null;
+        for (final File file : advancementDirectoryContent) {
+            if (file.getName().endsWith(".yml")) {
+                try {
+                    AdvancementTree tree = new AdvancementTree(file);
+                    if (file.getName().split(".yml")[0].contains(" ")) {
+                        CustomAdvancements.getInstance().getLogger().log(Level.SEVERE, "Advancement tree file name '" + file.getName() + "' contains spaces, this is not allowed.");
+                        continue;
+                    }
+                    advancementTrees.put(file.getName().split(".yml")[0], tree);
+                } catch (IllegalArgumentException e) {
+                    CustomAdvancements.getInstance().getLogger().log(Level.SEVERE, "Advancement tree has invalid name '" + file.getName() + "'. Make sure it only contains letters, numbers and underscores.");
+                }
+            }
+        }
+    }
 
-	/**
-	 * Executes the complete complete actions for an advancement and marks it as completed
-	 *
-	 * @param path       The path of an advancement formatted as 'treeLabel.advancementLabel'
-	 * @param playeruuid UUID of the minecraft player
-	 */
-	public void complete(final String path, final UUID playeruuid) throws InvalidAdvancementException {
-		val treeLabel = getAdvancementTreeLabel(path);
-		val advancementLabel = getAdvancementLabel(path);
-		getAdvancementTree(treeLabel).complete(advancementLabel,playeruuid);
-	}
+    /**
+     * Executes the complete actions for an advancement and marks it as completed
+     *
+     * @param path       The path of an advancement formatted as 'treeLabel.advancementLabel'
+     * @param playeruuid UUID of the minecraft player
+     */
+    public void complete(final String path, final UUID playeruuid) throws InvalidAdvancementException {
+        val treeLabel = getAdvancementTreeLabel(path);
+        val advancementLabel = getAdvancementLabel(path);
+        getAdvancementTree(treeLabel).complete(advancementLabel, playeruuid);
+    }
 
-	/**
-	 * Searches for the advancement type using the label it is registered with.
-	 * @param type the label of an advancement type
-	 * @return the {@link AdvancementType}
-	 */
-	public AdvancementType getAdvancementType(final String type) {
-		return advancementTypes.values().stream().filter(advancement -> advancement.equals(type)).findAny().orElseGet(Empty::new);
-	}
+    /**
+     * Searches for the advancement type using the label it is registered with.
+     *
+     * @param type the label of an advancement type
+     * @return the {@link AdvancementType}
+     */
+    public AdvancementType getAdvancementType(final String type) {
+        return advancementTypes.values().stream().filter(advancement -> advancement.equals(type)).findAny().orElseGet(Empty::new);
+    }
 
-	/**
-	 * Converts map {@link AdvancementManager#advancementTrees} into a list and returns it.
-	 * @return list of all registered {@link AdvancementTree}'s
-	 */
-	public List<AdvancementTree> getAdvancementTrees() {
-		return new ArrayList<>(advancementTrees.values());
-	}
+    /**
+     * Converts map {@link AdvancementManager#advancementTrees} into a list and returns it.
+     *
+     * @return list of all registered {@link AdvancementTree}'s
+     */
+    public List<AdvancementTree> getAdvancementTrees() {
+        return new ArrayList<>(advancementTrees.values());
+    }
 
-	/**
-	 * Searches {@link AdvancementManager#advancementTrees} for the tree matching the label
-	 *
-	 * @param label the unique label of an {@link AdvancementTree}
-	 * @return the {@link AdvancementTree} of the label given as input
-	 */
-	public AdvancementTree getAdvancementTree(final String label) throws InvalidAdvancementException {
-		if (advancementTrees.get(label) == null) throw new InvalidAdvancementException();
-		return advancementTrees.get(label);
-	}
+    /**
+     * Searches {@link AdvancementManager#advancementTrees} for the tree matching the label
+     *
+     * @param label the unique label of an {@link AdvancementTree}
+     * @return the {@link AdvancementTree} of the label given as input
+     */
+    public AdvancementTree getAdvancementTree(final String label) throws InvalidAdvancementException {
+        if (advancementTrees.get(label) == null) throw new InvalidAdvancementException();
+        return advancementTrees.get(label);
+    }
 
-	/**
-	 * Converts map {@link AdvancementManager#advancementTypes} into a list and returns it.
-	 * @return list of all registered {@link AdvancementType}'s
-	 */
-	public List<AdvancementType> getAdvancementTypes() {
-		return new ArrayList<>(advancementTypes.values());
-	}
+    /**
+     * Converts map {@link AdvancementManager#advancementTypes} into a list and returns it.
+     *
+     * @return list of all registered {@link AdvancementType}'s
+     */
+    public List<AdvancementType> getAdvancementTypes() {
+        return new ArrayList<>(advancementTypes.values());
+    }
 
-	/**
-	 * Gets the correct advancement from the correct tree registered.
-	 *
-	 * @param path The path of an advancement formatted as 'treeLabel.advancementLabel'
-	 * @return the {@link CAdvancement} belonging to that path
-	 */
-	public CAdvancement getAdvancement(final String path) throws InvalidAdvancementException {
-		val treeLabel = getAdvancementTreeLabel(path);
-		val advancementLabel = getAdvancementLabel(path);
-		if (treeLabel == null || advancementLabel == null) throw new InvalidAdvancementException();
-		AdvancementTree tree = getAdvancementTree(treeLabel);
-		if (tree == null) throw new InvalidAdvancementException();
-		return tree.getAdvancement(advancementLabel);
-	}
+    /**
+     * Gets the correct advancement from the correct tree registered.
+     *
+     * @param path The path of an advancement formatted as 'treeLabel.advancementLabel'
+     * @return the {@link CAdvancement} belonging to that path
+     */
+    public CAdvancement getAdvancement(final String path) throws InvalidAdvancementException {
+        val treeLabel = getAdvancementTreeLabel(path);
+        val advancementLabel = getAdvancementLabel(path);
+        if (treeLabel == null || advancementLabel == null) throw new InvalidAdvancementException();
+        AdvancementTree tree = getAdvancementTree(treeLabel);
+        if (tree == null) throw new InvalidAdvancementException();
+        return tree.getAdvancement(advancementLabel);
+    }
 
-	/**
-	 * Gets the tree label from the given path
-	 * @param path The path of an advancement formatted as 'treeLabel.advancementLabel'
-	 * @return the tree label of the given path
-	 */
-	public static String getAdvancementTreeLabel(final String path) {
-		return path.split("\\.")[0];
-	}
+    /**
+     * Gets the tree label from the given path
+     *
+     * @param path The path of an advancement formatted as 'treeLabel.advancementLabel'
+     * @return the tree label of the given path
+     */
+    public static String getAdvancementTreeLabel(final String path) {
+        return path.split("\\.")[0];
+    }
 
-	/**
-	 * Gets the advancement label from the given path
-	 *
-	 * @param path The path of an advancement formatted as 'treeLabel.advancementLabel'
-	 * @return the advancement label of the given path
-	 */
-	public static String getAdvancementLabel(final String path) {
-		return path.split("\\.")[1];
-	}
+    /**
+     * Gets the advancement label from the given path
+     *
+     * @param path The path of an advancement formatted as 'treeLabel.advancementLabel'
+     * @return the advancement label of the given path
+     */
+    public static String getAdvancementLabel(final String path) {
+        return path.split("\\.")[1];
+    }
 
-	/**
-	 * Searches for the advancement type using the label it is registered with.
-	 *
-	 * @param type the label of an advancement type
-	 * @return the {@link AdvancementType}
-	 */
-	public AdvancementRewardType getAdvancementRewardType(final String type) {
-		return advancementRewardTypes.values().stream().filter(advancement -> advancement.equals(type)).findAny().orElseGet(None::new);
-	}
+    /**
+     * Searches for the advancement type using the label it is registered with.
+     *
+     * @param type the label of an advancement type
+     * @return the {@link AdvancementType}
+     */
+    public AdvancementRewardType getAdvancementRewardType(final String type) {
+        return advancementRewardTypes.values().stream().filter(advancement -> advancement.equals(type)).findAny().orElseGet(None::new);
+    }
 
-	/**
-	 * Gets the type of an requirement using the label
-	 *
-	 * @param type label of an registered {@link AdvancementRequirementType}
-	 * @return the {@link AdvancementRequirementType} that belongs to the given label
-	 */
-	public AdvancementRequirementType getAdvancementRequirementType(final String type) {
-		return advancementRequirementTypes.values().stream().filter(advancement -> advancement.equals(type)).findAny().orElseGet(me.tippie.customadvancements.advancement.requirement.types.None::new);
-	}
+    /**
+     * Gets the type of an requirement using the label
+     *
+     * @param type label of an registered {@link AdvancementRequirementType}
+     * @return the {@link AdvancementRequirementType} that belongs to the given label
+     */
+    public AdvancementRequirementType getAdvancementRequirementType(final String type) {
+        return advancementRequirementTypes.values().stream().filter(advancement -> advancement.equals(type)).findAny().orElseGet(me.tippie.customadvancements.advancement.requirement.types.None::new);
+    }
 
-	/**
-	 * Unregisters all listeners, {@link AdvancementType}'s, {@link AdvancementTree}'s, {@link AdvancementRewardType}'s, {@link AdvancementRequirementType}'s
-	 */
-	public void unregisterAll() {
-		advancementTypes.values().forEach(HandlerList::unregisterAll);
+    /**
+     * Unregisters all listeners, {@link AdvancementType}'s, {@link AdvancementTree}'s, {@link AdvancementRewardType}'s, {@link AdvancementRequirementType}'s
+     */
+    public void unregisterAll() {
+        advancementTypes.values().forEach(HandlerList::unregisterAll);
 
-		advancementTypes.clear();
-		advancementTrees.clear();
-		advancementRewardTypes.clear();
-		advancementRequirementTypes.clear();
-	}
+        advancementTypes.clear();
+        advancementTrees.clear();
+        advancementRewardTypes.clear();
+        advancementRequirementTypes.clear();
+    }
 }

--- a/core/src/main/java/me/tippie/customadvancements/advancement/AdvancementTree.java
+++ b/core/src/main/java/me/tippie/customadvancements/advancement/AdvancementTree.java
@@ -6,6 +6,7 @@ import lombok.var;
 import me.tippie.customadvancements.CustomAdvancements;
 import me.tippie.customadvancements.advancement.requirement.AdvancementRequirement;
 import me.tippie.customadvancements.advancement.reward.AdvancementReward;
+import me.tippie.customadvancements.util.Utils;
 import org.bukkit.Bukkit;
 import org.bukkit.Material;
 import org.bukkit.configuration.ConfigurationSection;
@@ -49,6 +50,8 @@ public class AdvancementTree {
 	 */
 	AdvancementTree(final File config) {
 		label = config.getName().split(".yml")[0];
+		if (!Utils.validateNamespacedKey(label))
+			throw new IllegalArgumentException("Invalid advancement tree label: " + label);
 		CustomAdvancements.getInstance().getLogger().log(Level.INFO, "Attempting to load advancement tree " + config.getName());
 		try {
 			final FileConfiguration data = YamlConfiguration.loadConfiguration(config);
@@ -64,6 +67,11 @@ public class AdvancementTree {
 			}
 			assert treeAdvancements != null;
 			for (final String advancementLabel : treeAdvancements.getKeys(false)) {
+				if (!Utils.validateNamespacedKey(advancementLabel)){
+					CustomAdvancements.getInstance().getLogger().log(Level.SEVERE, "Advancement '" + advancementLabel + "' of tree '" + label + "' has an invalid label! Skipping advancement...");
+					continue;
+				}
+
 				String advancementType = treeAdvancements.getString(advancementLabel + ".type");
 				final String advancementValue = treeAdvancements.getString(advancementLabel + ".value");
 				int amount = treeAdvancements.getInt(advancementLabel + ".amount", -1);

--- a/core/src/main/java/me/tippie/customadvancements/util/Utils.java
+++ b/core/src/main/java/me/tippie/customadvancements/util/Utils.java
@@ -7,47 +7,12 @@ import org.bukkit.NamespacedKey;
 import org.bukkit.advancement.Advancement;
 import org.bukkit.entity.Player;
 
+import java.util.regex.Pattern;
+
 public class Utils {
+    private final static Pattern NAMESPACED_KEY_PATTERN = Pattern.compile("^[a-z0-9/._-]+$");
 
-    // Format for JSON component:
-    // https://minecraft.fandom.com/wiki/Advancement/JSON_format
-    public static void showToast(Player player, String title, String description){
-        JsonObject advancementJSON = new JsonObject();
-
-        JsonObject display = new JsonObject();
-        display.addProperty("show_toast", true);
-        display.addProperty("announce_to_chat", false);
-        display.addProperty("hidden",true);
-        display.addProperty("frame", "task");
-        display.addProperty("title", title);
-        display.addProperty("description", description);
-
-        JsonObject icon = new JsonObject();
-        icon.addProperty("item", "minecraft:stone");
-        display.add("icon", icon);
-
-        //Creating an impossible critirion to grant it using the plugin
-        JsonObject criteria = new JsonObject();
-        JsonObject crit = new JsonObject();
-
-        crit.addProperty("trigger","minecraft:bee_nest_destroyed");
-        JsonObject conditions = new JsonObject();
-        conditions.addProperty("block", "minecraft:bedrock");
-        JsonObject item = new JsonObject();
-        item.addProperty("item", "minecraft:air");
-        conditions.add("item", item);
-        conditions.addProperty("num_bees_inside", 100);
-        criteria.add("crit",crit);
-        advancementJSON.add("criteria", criteria);
-
-
-        advancementJSON.add("display",display);
-        NamespacedKey key = new NamespacedKey(CustomAdvancements.getInstance(), "showtoast"+Math.round(Math.random()*10000));
-        Advancement advancement = Bukkit.getUnsafe().loadAdvancement(key, String.valueOf(advancementJSON));
-        for (String critirion : player.getAdvancementProgress(advancement).getRemainingCriteria()){
-            player.getAdvancementProgress(advancement).awardCriteria(critirion);
-        }
-
-        Bukkit.getUnsafe().removeAdvancement(key);
+    public static boolean validateNamespacedKey(String str) {
+        return NAMESPACED_KEY_PATTERN.matcher(str).matches();
     }
 }

